File: frontend/src/services/table-mapping-service.js
import axios from 'axios';
import authHeader from './auth-header';
import {environment} from "./environment";

const API_URL = environment.TABLE_MAP_URL;

class TableMappingService {

    /**
     * Sends a GET request to API to retrieve data from a table mapping (field of each table and field mappings)
     *
     * @param tableMappingId table mapping id
     * @param etl_id ETL procedure's id
     * @returns mapping table mapping information
     */

    getMapping(tableMappingId, etl_id) {
        const username = JSON.parse(localStorage.getItem('user')).username;
        return axios.get(
            API_URL + "map/" + tableMappingId,
            { headers: authHeader(), params: { etl_id: etl_id, username: username }}
        );
    }


    /**
     * Sends a POST request to API to create a new table mapping
     * 
     * @param {*} etl_id ETL procedure id
     * @param {*} ehrTableId source table id
     * @param {*} omopTableId target table id
     * @returns table mapping created
     */

    addTableMapping = async (etl_id, ehrTableId, omopTableId) => {
        const username = JSON.parse(localStorage.getItem('user')).username;
        return await axios.post(
            API_URL + "map", 
            {}, 
            {
                headers: authHeader(),
                params: {
                    ehrTableId: ehrTableId,
                    omopTableId: omopTableId,
                    username: username,
                    etl_id: etl_id
                }
            }
        );
    }


    /**
     * Sends a DELETE request to delete table mapping
     * 
     * @param etl_id ETL procedure id
     * @param tableMappingId table mapping id
     * @returns ETL procedure
     */

    removeTableMapping(etl_id, tableMappingId) {
        const username = JSON.parse(localStorage.getItem('user')).username;
        return axios.delete(
            API_URL + "map",
            { headers: authHeader(), params: { tableMappingId: tableMappingId, etl_id: etl_id, username:username } }
        )
    }


    /**
     * Sends a PUT request to change table mapping completion status
     *
     * @param tableMapping table mapping id
     * @param complete completion status
     * @param etl_id ETL procedure's id
     * @returns table mapping edited
     */

    editCompleteMapping(tableMapping, complete, etl_id) {
        const username = JSON.parse(localStorage.getItem('user')).username;
        return axios.put(
            API_URL + "map/" + tableMapping + "/complete",
            null,
            { headers: authHeader(), params: { completion: complete, etl_id: etl_id, username: username }}
        );
    }


    /**
     * Sends a PUT request to change table mapping logic
     * 
     * @param map_id table mapping id
     * @param logic mapping logic
     * @param etl_id ETL procedure's id
     * @returns table mapping edited
     */

    editMappingLogic(map_id, logic, etl_id) {
        const username = JSON.parse(localStorage.getItem('user')).username;
        return axios.put(
            API_URL + "map/" + map_id + "/logic",
            null,
            { headers: authHeader(), params: { logic: logic, etl_id: etl_id, username: username }}
        )
    }

}

export default new TableMappingService();

File: backend/riaw/src/main/java/com/ua/riaw/service/tableMapping/TableMappingService.java
package com.ua.riaw.service.tableMapping;

import com.ua.riaw.model.CDMVersion;
import com.ua.riaw.model.ETL;
import com.ua.riaw.model.TableMapping;
import com.ua.riaw.model.ehr.EHRDatabase;
import com.ua.riaw.model.ehr.EHRTable;
import com.ua.riaw.model.omop.OMOPDatabase;
import com.ua.riaw.model.omop.OMOPTable;

import java.util.List;

public interface TableMappingService {

    /**
     * Gets a table mapping given its id
     *
     * @param map_id table mapping id
     * @param etl_id ETL procedure's id
     * @param username user's username
     * @return table mapping if found, null otherwise
     */

    TableMapping getTableMappingById(Long map_id, Long etl_id, String username);


    TableMapping getTableMappingById(Long map_id);

    /**
     * Creates a table mapping between a table from the EHR database and a table from the OMOP CDM
     *
     * @param ehrTableId source table's id
     * @param omopTableId target table's id
     * @param etl_id ETL procedure's id
     * @param username User's username
     * @return created table mapping
     */

    TableMapping addTableMapping(Long ehrTableId, Long omopTableId, Long etl_id, String username);


    /**
     * Deletes a table mapping given its id
     *
     * @param tableMappingId table mapping id
     * @param etl_id ETL procedure's id
     * @param username User's username
     */

    void removeTableMapping(Long tableMappingId, Long etl_id, String username);


    /**
     * Changes the completion state of a table mapping
     *
     * @param tableMappingId table mapping id
     * @param completion state to change to
     * @param etl_id ETL procedure's id
     * @param username User's username
     * @return altered table mapping
     */

    TableMapping changeCompletionStatus(Long tableMappingId, boolean completion, Long etl_id, String username);


    /**
     * Changes the table mapping logic
     *
     * @param tableMappingId table mapping id
     * @param logic logic to change to
     * @param etl_id ETL procedure's id
     * @param username User's username
     * @return altered table mapping
     */

    TableMapping changeMappingLogic(Long tableMappingId, String logic, Long etl_id, String username);


    /**
     * Creates the table mapping contained in a JSON file
     *
     * @param etl ETL procedure object
     * @param tableMappings table mapping in JSON
     * @param ehrDatabase source database
     * @param omopDatabase target database
     * @return table mappings created
     */

    List<TableMapping> getTableMappingsFromJSON(ETL etl, List<TableMapping> tableMappings, EHRDatabase ehrDatabase, OMOPDatabase omopDatabase);


    /**
     * Removes all table mappings of a given ETL procedures
     *
     * @param etl_id ETL procedure's id
     */

    void removeTableMappingsFromETL(long etl_id);


    /**
     * Creates mapping to or from a stem table (stored in file)
     *
     * @param version OMOP CDM version
     * @param omopDatabase target database
     * @param sourceStemTable stem table on EHR database
     * @param etl ETL procedure object
     * @return list of created table mappings
     */

    List<TableMapping> createMappingsWithStemTable(CDMVersion version, OMOPDatabase omopDatabase, EHRTable sourceStemTable, ETL etl);




























    /**
     * Gets all table mappings from a given ETL procedure
     *
     * @param etl_id ETL procedure's id
     * @return list with table mappings
     */

    List<TableMapping> getTableMappingFromETL(Long etl_id);


    void removeTableMappingsFromTable(Long etl_id, EHRTable table);

    void removeTableMappingsToTable(Long etl_id, OMOPTable table);
}
